lib/src/server.dart

```diff
@@ -0,0 +1,85 @@
+import 'dart:convert';
+import 'package:meta/meta.dart';
+import 'package:xml/xml.dart';
+
+import 'common.dart';
+import 'converter.dart';
+
+/// A [XmlRpcHandler] handles the handling RPC functions along with marshalling the arguments and results to / from XMLRPC spec
+///
+/// Has a set of [codecs] for encoding and decoding the datatypes
+/// To register a method or function with the dispatcher call [registerFunction] along with the function and name
+class XmlRpcHandler {
+  /// The [codecs] used for encoding and decoding
+  List<Codec> codecs;
+
+  /// The function registry
+  final Map<String, Function> methods;
+
+  /// Creates a [XmlRpcHandler] with the set of [codecs] for encoding and decoding
+  XmlRpcHandler({@required this.methods, this.codecs}) {
+    codecs ??= standardCodecs;
+  }
+
+  /// Marshalls the [data] from XML to Dart types, and then dispatches the function, and marshals the return value back into the XMLRPC format
+  XmlDocument handle(XmlDocument document) {
+    final methodCall = document.findElements('methodCall').first;
+    final methodName = methodCall.findElements('methodName').first.text;
+    final parsedArgs = [];
+    try {
+      final args =
+          methodCall.findElements('params').first.findElements('param');
+      for (final arg in args) {
+        parsedArgs.add(
+          decode(getValueContent(arg.findElements('value').first), codecs),
+        );
+      }
+    } on StateError {
+      // No arguments given
+    }
+    final returnValue = _dispatch(methodName, parsedArgs);
```

Comment:
We can avoid try-catch with null-safe operators.

```suggestion
    final args = methodCall
        .findElements('params')
        ?.first
        ?.findElements('param')
        ?.map((param) => param.findElements('value').first)
        ?.map(getValueContent)
        ?.map((e) => decode(e, codecs))
        ?.toList();
    final returnValue = await Function.apply(methods[methodName], args ?? []);
```

Reply:
If there are no params, `methodCall.findElements('params')` does not return `null`, but the getter `first` throws a `StateError`. I would need to check if `methodCall.findElements('params').isEmpty()` which doesn't fit into the null cascade as easily. But I think I can do something similar, and avoid the try / catch.